### PR TITLE
credential-reference changes with regards to mail subsystem integration

### PR DESCRIPTION
**Pull request by pskopek**
https://issues.jboss.org/browse/WFCORE-2210
https://issues.jboss.org/browse/WFCORE-2211
Fixed issues of credential-reference marshalling using standard marshaller.
Removed Destroyable interface implementation as it is not needed any more.

### CODE REVIEW

**Comment by bstansberry** [comment]
Please clone JBEAPs into WFCORE before working them.

File: controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java [hunk 1]
```diff
@@ -169,12 +169,21 @@ private static AttributeMarshaller credentialReferenceAttributeMarshaller() {
...
+                    crNode = credentialReferenceModelNode.get(attribute.getName());
                 } else {
-                    credentialStoreAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
-                    credentialAliasAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
-                    credentialTypeAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
+                    crNode = credentialReferenceModelNode;
+                }
+                if (!crNode.isDefined()) {
+                    throw new XMLStreamException(ControllerLogger.ROOT_LOGGER.invalid(attribute.getName()));
+                }
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java]: If this is possible something has been allowed previously that should have been detected and rejected. We shouldn't be doing model validation during xml marshaling.

If this shouldn't be possible and this is the equivalent of an assert, perhaps use an assert.

File: controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java [hunk 2]
```diff
@@ -169,12 +168,21 @@ private static AttributeMarshaller credentialReferenceAttributeMarshaller() {
...
-                if (credentialReferenceModelNode.hasDefined(clearTextAttribute.getName())) {
-                    clearTextAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
+                ModelNode crNode = null;
+                if (credentialReferenceModelNode.hasDefined(attribute.getName())) {
+                    crNode = credentialReferenceModelNode.get(attribute.getName());
                 } else {
-                    credentialStoreAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
-                    credentialAliasAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
-                    credentialTypeAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
+                    crNode = credentialReferenceModelNode;
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java]: The existence of this 'else' block and the name of this 'credentialReferenceModelNode' param tells me this has gone wrong. The ModelNode param is meant to be a ModelNode that is a parent, with this attribute as a child. Generally that means it is a ModelNode that represents the Resource, with children for all attributes in that resource. The param is not meant to be the attribute value itself. If it ever is I fear something has gone wrong.

Have a look at DefaultAttributeMarshaller for the standard kind of implementation. A check of isMarshallable() with no xml writing at all if that returns false, and then the attribute data is extracted from the model as you do here in L173.

Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java]: Note I have assigned a JIRA to @ctomc to polish up the javadoc on AttributeMarshaller. The marshallAsElement javadoc is missing, although marshalAsAttribute gives a good indication of the expected behavior.

**Comment by pskopek** [comment]
@bstansberry WFCORE JIRAs created and referenced
and marshaller and parser set to standard ones.
Thanks for spotting it!

File: controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java [hunk 3]
```diff
@@ -135,72 +129,30 @@ public static ObjectTypeAttributeDefinition getAttributeDefinition() {
...
-                return namedNode.asString();
-            }
-            return null;
+    public static String credentialReferencePartAsStringIfDefined(ModelNode credentialReferenceValue, String name) throws OperationFailedException {
+        assert !credentialReferenceValue.isDefined() : credentialReferenceValue;
+        ModelNode result = credentialReferenceValue.get(name);
+        if (result.isDefined()) {
+            return result.asString();
         }
         return null;
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java]: Something seems odd here.

L148 asserts that credentialReferenceValue is undefined. So the test at L150 will always test false and the method will return null.

On L148 did you not intend for the '!' to be there?

Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java]: Yeah, reading the usage I'm almost sure the '!' was by mistake. :)

Comment (pskopek) [inline controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java]: yes, wrong copy/paste. Fixed.

**Pull request merged**